File: test_resources/kotlin/enum_with_positions.kt
sealed interface MyEnum {
    data object Idle : MyEnum

    data class Move(
        val y: Double,
        val x: Double,
    ) : MyEnum

    data class Update(
        val p1: String,
        val p2: Float,
        val p3: Int,
        val p8: Double,
    ) : MyEnum

    companion object {
        fun createDefault(): MyEnum = Idle

        fun readFromBuffers(reader: Long): MyEnum {
            return when (val case = UInt.readFromBuffers(reader)) {
                3U -> Idle
                1U -> {
                    val y = Double.readFromBuffers(reader)
                    val x = Double.readFromBuffers(reader)

                    Move(
                        y = y,
                        x = x,
                    )
                }
                2U -> {
                    val p1 = String.readFromBuffers(reader)
                    val p2 = Float.readFromBuffers(reader)
                    val p3 = Int.readFromBuffers(reader)
                    val p8 = Double.readFromBuffers(reader)

                    Update(
                        p1 = p1,
                        p2 = p2,
                        p3 = p3,
                        p8 = p8,
                    )
                }
                else -> throw IllegalArgumentException("Invalid enum value: $case")
            }
        }

        fun skipInBuffers(reader: Long, count: Int) {
            for (i in 0..<count) {
                readFromBuffers(reader)
            }
        }
    }

    fun writeToBuffers(writer: Long) {
        when (this) {
            Idle -> {
                3UL.writeToBuffers(writer)
            }
            is Move -> {
                1UL.writeToBuffers(writer)
                y.writeToBuffers(writer)
                x.writeToBuffers(writer)
            }
            is Update -> {
                2UL.writeToBuffers(writer)
                p1.writeToBuffers(writer)
                p2.writeToBuffers(writer)
                p3.writeToBuffers(writer)
                p8.writeToBuffers(writer)
            }
        }
    }
}

File: test_resources/swift/enum_with_positions.swift
enum MyEnum: TechPawsBuffersModel {
    case idle
    case move(
        /* y */ Double,
        /* x */ Double
    )
    case update(
        String,
        Float,
        Int32,
        Double
    )

    static func createBuffersDefault() -> MyEnum {
        return .idle
    }

    static func readFromBuffers(
        _ bytesReader: TechPawsBuffersBytesReader
    ) -> Self {
        let caseValue = bytesReader.readUInt32()

        switch caseValue {
        case 3:
            return .idle
        case 1:
            let y = bytesReader.readDouble()
            let x = bytesReader.readDouble()

            return .move(
                y,
                x
            )
        case 2:
            let p1 = String.readFromBuffers(bytesReader)
            let p2 = bytesReader.readFloat()
            let p3 = bytesReader.readInt32()
            let p8 = bytesReader.readDouble()

            return .update(
                p1,
                p2,
                p3,
                p8
            )
        default:
            fatalError("Invalid value: \(caseValue)")
        }
    }

    static func skipInBuffers(
        _ bytesReader: TechPawsBuffersBytesReader,
        _ count: UInt64
    ) {
        for _ in 1...count {
            let caseValue = bytesReader.readUInt32()

            switch caseValue {
            case 3:
                continue
            case 1:
                let _ = bytesReader.readDouble()
                let _ = bytesReader.readDouble()
            case 2:
                let _ = String.readFromBuffers(bytesReader)
                let _ = bytesReader.readFloat()
                let _ = bytesReader.readInt32()
                let _ = bytesReader.readDouble()
            default:
                fatalError("Invalid value: \(caseValue)")
            }
        }
    }

    func writeToBuffers(
        _ bytesWriter: TechPawsBuffersBytesWriter
    ) {

        switch self {
        case .idle:
            bytesWriter.writeUInt32(3)
        case .move(
            let y,
            let x
        ):
            bytesWriter.writeUInt32(1)
            bytesWriter.writeDouble(y)
            bytesWriter.writeDouble(x)
        case .update(
            let p1,
            let p2,
            let p3,
            let p8
        ):
            bytesWriter.writeUInt32(2)
            p1.writeToBuffers(bytesWriter)
            bytesWriter.writeFloat(p2)
            bytesWriter.writeInt32(p3)
            bytesWriter.writeDouble(p8)
        }
    }
}
